osu.Game/Context/IContext.cs

```diff
@@ -0,0 +1,13 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+namespace osu.Game.Context;
+
+public interface IContext
```

Comment:
All of this `IContext` machinery needs better naming and xmldoc at the least, or preferably not exist at all. I have no idea what anything is supposed to be here. It looks to be a generic "attach anything to everything" mechanism for whose existence I see no real reason. Is there one? Is this taking some particular far-off direction that I'm failing to notice?

My questions would be in order:

- Does this PR need to touch difficulty points at all, and if so, then why? I can understand wanting get rid of sample points, because of https://github.com/ppy/osu/issues/15618#issuecomment-1224120633 and https://github.com/ppy/osu/issues/22171, but I don't see much point or reasoning in doing this for difficulty points.
- As far as I can see, right now this only serves the purpose of attaching the "legacy bpm multiplier" stuff. Could that issue be solved in a different way that does not require this baroqueness?

Reply:
- Touching the difficulty points came from the same idea as removing sample control points from `HitObject', namely "Why have any 'control point' stored in `HitObject` if it only functions as a container for some properties of the `HitObject`?". And difficulty control points were particularly weird, they store a slider velocity for all hit objects even though that's only relevant for sliders, and the legacy information just seems tacked on just because it was possible. Since I'm already changing the whole data flow, might as well fix that too.
- I gotta admit the `IContext` stuff is quite experimental. Removing it would result in directly adding the legacy bpm multiplier stuff as a property of `HitObject` and I'd need some other way of passing the information about `GenerateTicks` from the legacy decoder to the `Slider` class. The `IContext` system made it possible to attach this information to a `HitObject` without cluttering it with legacy properties.

Reply:
Just to clarify: don't misunderstand me here, I'm completely fine with `IHasSliderVelocity` et al. I'm much less fine with the API of "attach anything you like here", because if we open that Pandora box, there's no telling what ruleset authors are going to go and do with it.

All I'm saying is that I'd want to have a more constrained interface here. And what I'd be okay is probably a middle ground of what you have here and what we have in `master` - namely, the interface still only applies to things where slider velocity makes sense, but the member exposed by the interface would be essentially what is currently the difficulty control point.

Reply:
I've removed `IContext` and added an interface `IHasGenerateTicks` to allow storing the `GenerateTicks` property in `Slider` and I added the `LegacyBpmMultiplier` to `HitObject`.